### PR TITLE
Added 5 modules to Design Patterns Java Series

### PR DESCRIPTION
**Pull request by teo-nikolov**

### CODE REVIEW

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategyadvanced/core/BasePage.java [hunk 1]
```diff
@@ -1,6 +1,6 @@
 package strategyadvanced.core;
 
-import java.lang.reflect.ParameterizedType;
+import pageobjectadvanced.core.ReflectionNewInstanceFactory;
```
Comment (teo-nikolov) [inline DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategyadvanced/core/BasePage.java:3]: should import **strategyadvanced.core.ReflectionNewInstanceFactory** instead, will fix in next commit

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/decorator/services/CouponCodeCalculationService.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package decorator.services;
+
+import decorator.enums.CouponCode;
+
+public class CouponCodeCalculationService {
+    double discount;
+    public double calculate(double price, String couponCode) {
```
Comment (angelovstanton) [inline DesignPatternsInAutomatedTestingJava-Series/src/test/java/decorator/services/CouponCodeCalculationService.java:7]: leave empty line

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/decorator/services/VatTaxCalculationService.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package decorator.services;
+
+import decorator.enums.Country;
+
+public class VatTaxCalculationService {
+    double taxValue;
+    public double calculate(double price, Country country) {
```
Comment (angelovstanton) [inline DesignPatternsInAutomatedTestingJava-Series/src/test/java/decorator/services/VatTaxCalculationService.java:7]: leave empty line

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/pageobjectadvanced/core/Driver.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    public static void startBrowser(BrowserType browserType, int defaultTimeout) {
+
+        switch (browserType) {
+            case FIREFOX:
+                WebDriverManager.firefoxdriver().setup();
+                setBrowser(new FirefoxDriver());
+                break;
+            case CHROME:
+                //setBrowser(new ChromeDriver());
```
Comment (angelovstanton) [inline DesignPatternsInAutomatedTestingJava-Series/src/test/java/pageobjectadvanced/core/Driver.java:42]: uncomment this everywhere and make it work :)

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategy/core/BasePage.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    public void navigate(String part) {
+        Driver.getBrowser().navigate().to(url.concat(part));
+    }
+
+    public void navigate() {
+        Driver.getBrowser().navigate().to(url);
+    }
+
+    @SuppressWarnings("unchecked")
```
Comment (angelovstanton) [inline DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategy/core/BasePage.java:23]: even if the methods are with different access modifiers  I believe it is more readable elements and assertions to be placed together one after another. I added similar note I think to the advanced page object model article. This follows Oracle Java convention and it is OK.

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategy/core/Driver.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        getBrowser().quit();
+        setBrowser(null);
+        setBrowserWait(null);
+    }
+
+    public static void waitForAjax() {
+        var javascriptExecutor = (JavascriptExecutor) browser;
+        browserWait.until(d -> javascriptExecutor.executeScript("return window.jQuery != undefined && jQuery.active == 0"));
+        try {
+            Thread.sleep(1000);
```
Comment (angelovstanton) [inline DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategy/core/Driver.java:73]: no need

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategy/core/ReflectionNewInstanceFactory.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package strategy.core;
+
+import java.lang.reflect.ParameterizedType;
+
+public class ReflectionNewInstanceFactory {
+    @SuppressWarnings("unchecked, rawtypes")
+    public static Object getTypeParameter(int index, Class getClass) {
+        try {
+            var elementsClass = (Class) ((ParameterizedType) getClass
```
Comment (angelovstanton) [inline DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategy/core/ReflectionNewInstanceFactory.java:9]: remove spaces, I commented it above

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategy/pages/itempage/ItemAssertions.java [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package strategy.pages.itempage;
+
+import strategy.core.BaseAssertions;
+
+public class ItemAssertions extends BaseAssertions<ItemElements> {}
```
Comment (angelovstanton) [inline DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategy/pages/itempage/ItemAssertions.java:5]: do you need empty assertions class?

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategyadvanced/StorePurchaseAdvancedStrategyTests.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        purchaseInfo.setCompany("Space Flowers");
+        purchaseInfo.setCountry("Germany");
+        purchaseInfo.setAddress1("1 Willi Brandt Avenue Tiergarten");
+        purchaseInfo.setAddress2("Lьtzowplatz 17");
+        purchaseInfo.setCity("Berlin");
+        purchaseInfo.setZip("10115");
+        purchaseInfo.setPhone("+491888999281");
+        purchaseInfo.setCouponCode("happybirthday");
+
+        new PurchaseContext(new VatTaxOrderPurchaseStrategy(), new CouponCodeOrderPurchaseStrategy())
```
Comment (angelovstanton) [inline DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategyadvanced/StorePurchaseAdvancedStrategyTests.java:47]: purchaseItem

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategyadvanced/core/Driver.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        getBrowser().quit();
+        setBrowser(null);
+        setBrowserWait(null);
+    }
+
+    public static void waitForAjax() {
+        var javascriptExecutor = (JavascriptExecutor) browser;
+        browserWait.until(d -> javascriptExecutor.executeScript("return window.jQuery != undefined && jQuery.active == 0"));
+        try {
+            Thread.sleep(1000);
```
Comment (angelovstanton) [inline DesignPatternsInAutomatedTestingJava-Series/src/test/java/strategyadvanced/core/Driver.java:73]: no need

**Review by angelovstanton — CHANGES_REQUESTED** [review]

File: DesignPatternsInAutomatedTestingJava-Series/src/test/java/decorator/core/Driver.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2021 Automate The Planet Ltd.
+ * Author: Anton Angelov
```
Comment (angelovstanton) [inline DesignPatternsInAutomatedTestingJava-Series/src/test/java/decorator/core/Driver.java:3]: you can set your name here :)

**Review by angelovstanton — APPROVED** [review]

**Pull request merged**